src/native/http_request_utils.c

```diff
@@ -177,10 +178,86 @@ struct aws_input_stream *aws_input_stream_new_from_java_http_request_body_stream
     return NULL;
 }
 
+static inline int s_marshal_http_header_to_buffer(
+    struct aws_byte_buf *buf,
+    const struct aws_byte_cursor *name,
+    const struct aws_byte_cursor *value) {
+    if (aws_byte_buf_reserve_relative(buf, sizeof(int) + sizeof(int) + name->len + value->len)) {
+        return AWS_OP_ERR;
+    }
+
+    aws_byte_buf_write_be32(buf, (uint32_t)name->len);
+    aws_byte_buf_write_from_whole_cursor(buf, *name);
+    aws_byte_buf_write_be32(buf, (uint32_t)value->len);
+    aws_byte_buf_write_from_whole_cursor(buf, *value);
+    return AWS_OP_SUCCESS;
+}
+
+int aws_marshall_http_headers_to_dynamic_buffer(
+    struct aws_byte_buf *buf,
+    const struct aws_http_header *header_array,
+    size_t num_headers) {
+    for (size_t i = 0; i < num_headers; ++i) {
+        if (s_marshal_http_header_to_buffer(buf, &header_array[i].name, &header_array[i].value)) {
+            return AWS_OP_ERR;
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static inline int s_unmarshall_http_request(struct aws_http_message *message, struct aws_byte_cursor *request_blob) {
+    uint32_t field_len = 0;
+
+    if (!aws_byte_cursor_read_be32(request_blob, &field_len)) {
+        return AWS_OP_ERR;
```

Comment:
need to raise an error if the aws_byte_cursor_read_XYZ() functions return false

Reply:
yikes, this is why we probably shouldn't have had bool apis. good catch.